### PR TITLE
Ignore documents with empty ids in changes feed

### PR DESCRIPTION
**Pull request by rhyshort**
Ignore documents with an empty id in the changes feed. This
allows replication with databases which contain documents
with empty id fields as a result in a bug in CouchDB.

BugzID 24255

reviewer @mikerhodes 
reviewer @alfinkel

### CODE REVIEW

**Comment by alfinkel** [comment]
looks good to me.  +1.

File: sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java [hunk 1]
```diff
@@ -147,6 +164,58 @@ public void call_coordinationThreadInterrupted_replicatorStopCorrectly() throws
...
+        pullStrategy.run();
+
+        //should have zero documents
+        Assert.assertEquals(this.datastore.getDocumentCount(),0);
+        //make sure the correct events were fired
+        verify(mockListener).complete(any(ReplicationStrategyCompleted.class));
+        verify(mockListener,never()).error(any(ReplicationStrategyErrored.class));
+
+
+    }
```
Comment (mikerhodes) [inline sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java]: How difficult would it be to add a test where there are some correct entries (to make sure that the presence of an empty ID parameter doesn't break replication of good change entries)? Right now you're not checking that empty IDs are skipped rather than just the whole thing failing (perhaps silently).

Comment (rhyshort) [inline sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java]: it shouldn't be too difficult, it'll just need a couple of fixtures for openRevs,  add more documents into the changes feed fixture and a few more mocking of method calls

Comment (mikerhodes) [inline sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java]: It'd be good if we could do that, it'll catch more future regressions.

Comment (rhyshort) [inline sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java]: I have a working version with one empty doc id and one with a valid doc id. Do you think 2 docs for the test are fine or should I add more?

Comment (mikerhodes) [inline sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java]: Two docs works, but make sure the valid doc ID is *after* the dud one. Ideally you'd have one either side.

Comment (rhyshort) [inline sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java]: Added two doc test in 71c89e2 empty docId comes first in the mocked changes feed

File: fixture/empty_docid.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        "seq": 28,
+        "id": "4d3b3f01362649d79b31d9092799a7e0",
+        "changes": [
+          {
+            "rev": "1-13d33701a0954729ad029adf8fdc5a04"
+          }
+        ]
+      }
+     ]
+ }
```
Comment (mikerhodes) [inline fixture/empty_docid.json]: Format the two results in the changes feed the same way.

File: fixture/empty_docid_open_revs.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+[
+{"ok":{"_id":"","_rev":"1-bd42b942b8b672f0289cf3cd1f67044c","hello":"foo",
+  "_revisions": {
+  "start": 1,
+  "ids": ["bd42b942b8b672f0289cf3cd1f67044c"]
+}}}
+]
```
Comment (mikerhodes) [inline fixture/empty_docid_open_revs.json]: Format decently, add the newline at end of file.

File: fixture/empty_docid_test_open_revs_2.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+[
+  {"ok":{"_id":"4d3b3f01362649d79b31d9092799a7e0","_rev":"1-13d33701a0954729ad029adf8fdc5a04","hello":"world","foo":"bar",
+    "_revisions": {
+      "start": 1,
+      "ids": ["13d33701a0954729ad029adf8fdc5a04"]
+    }
+  }}
+]
```
Comment (mikerhodes) [inline fixture/empty_docid_test_open_revs_2.json]: Again, decent formatting.

File: sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java [hunk 5]
```diff
@@ -147,6 +164,56 @@ public void call_coordinationThreadInterrupted_replicatorStopCorrectly() throws
...
+
+            }
+        });
+        revs = new ArrayList<String>();
+        revs.add("1-13d33701a0954729ad029adf8fdc5a04");
+        when(mockRemoteDb.getRevisions("4d3b3f01362649d79b31d9092799a7e0", revs, new HashSet<String>(),false)).then(new Answer<Object>() {
+            @Override
+            public Object answer(InvocationOnMock invocation) throws Throwable {
+
+                return getMockedRevs("fixture/empty_docid_test_open_revs_2.json");
```
Comment (mikerhodes) [inline sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java]: Make the file names consistent, probably also prefix with the test name?

```
testReplicationDocWithEmptyId_changes.json
testReplicationDocWithEmptyId_open_revs_1.json
testReplicationDocWithEmptyId_open_revs_2.json
```

File: sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java [hunk 6]
```diff
@@ -147,6 +164,56 @@ public void call_coordinationThreadInterrupted_replicatorStopCorrectly() throws
...
+
+            }
+        });
+
+        //when(remoteDb.)
```
Comment (mikerhodes) [inline sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java]: Delete?

File: sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java [hunk 7]
```diff
@@ -147,6 +164,56 @@ public void call_coordinationThreadInterrupted_replicatorStopCorrectly() throws
...
+
+        StrategyListener mockListener = mock(StrategyListener.class);
+        final BasicPullStrategy pullStrategy = new BasicPullStrategy(createPullReplication());
+        pullStrategy.sourceDb = mockRemoteDb;
+        pullStrategy.getEventBus().register(mockListener);
+        pullStrategy.run();
+
+        //should have zero documents
+        Assert.assertEquals(this.datastore.getDocumentCount(), 1);
```
Comment (mikerhodes) [inline sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java]: Comment doesn't match assert.

File: sync-core/src/main/java/com/cloudant/sync/replication/BasicPullStrategy.java [hunk 8]
```diff
@@ -387,6 +387,10 @@ private ChangesResultWrapper nextBatch() throws DatastoreException {
 
         List<Callable<DocumentRevsList>> tasks = new ArrayList<Callable<DocumentRevsList>>();
         for(String id : ids) {
+            //skip any document with an empty id
+            if(id.isEmpty()){
+                continue;
+            }
```
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/replication/BasicPullStrategy.java]: Possibly output a debug log message here to say that we saw a changes feed entry with an empty ID and are ignoring it.

File: sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java [hunk 9]
```diff
@@ -147,6 +164,56 @@ public void call_coordinationThreadInterrupted_replicatorStopCorrectly() throws
...
+        //make sure the correct events were fired
+        verify(mockListener).complete(any(ReplicationStrategyCompleted.class));
+        verify(mockListener,never()).error(any(ReplicationStrategyErrored.class));
+
+
```
Comment (mikerhodes) [inline sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java]: Clean up the newlines.

File: sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java [hunk 10]
```diff
@@ -159,4 +226,22 @@ public void error(ReplicationStrategyErrored re) {
         }
     }
 
+    private List<DocumentRevs> getMockedRevs(String fixturePath) throws Exception{
```
Comment (mikerhodes) [inline sync-core/src/test/java/com/cloudant/sync/replication/BasicPullStrategyMockTest.java]: Not a very descriptive name, suggest calling it `loadOpenRevsResponseFromFixture`.

**Comment by mikerhodes** [comment]
:+1:

**Comment by mikerhodes** [comment]
Is this a bug in iOS too?

**Comment by rhyshort** [comment]
I am unaware of it, I'll raise an FB ticket to investigate
